Minimixxx controller config

ywwg opened the pull request:
There is currently no documentation for this config, but I will work on that. Seeing as I am the only person on Earth with this controller that seems ok for now :)

ywwg committed:
MiniMixxx: Initial implementation

ywwg committed:
MiniMixxx: Start implementing buttons

ywwg committed:
Merge branch 'main' into minimixxx

ywwg committed:
MiniMixxx: Implement layers

ywwg committed:
MiniMixxx: relative pitch slider mode

ywwg committed:
MiniMixxx: wheel spinnies!

ywwg committed:
MiniMixxx: forgot to hook up a connection

ywwg committed:
MiniMixxx: small light tweak

ywwg committed:
MiniMixxx: sampler layer

ywwg committed:
MiniMixxx: Fix some layer switching issues

ywwg committed:
MiniMixxx: library mode

ywwg committed:
MiniMixxx: add fx layer

oh no I have no room for hotcues!

ywwg committed:
MiniMixxx: rearrange file

ywwg committed:
MiniMixxx: factor out vu meter color

ywwg committed:
MiniMixxx: Add hotcue layer

Because we are out of buttons, hotcues are activated for both decks at once.  That seems ok.

MiniMixxx is now feature complete!

ywwg committed:
MiniMixxx: allow for scrubbing through track in library mode

ywwg committed:
MiniMixxx: Add newline to satisfy linter

ywwg committed:
MiniMixxx: Factor out colors

ywwg committed:
MiniMixxx: Add main output adjustments

ywwg committed:
MiniMixxx: trying to get consistent behavior with layers

ywwg committed:
MiniMixxx: cleanup

ywwg committed:
MiniMixxx: more cleanup

Holzhaus commented:
> Seeing as I am the only person on Earth with this controller that seems ok for now :)

If there is zero chance that anyone else has this controller, does it really make sense to add it to the mixxx source tree and ship it to every single user?

Why not just keep it in your personal mixxx config dir? Or are you planning mass-production soon? 🙂

Swiftb0y commented:
I don't think it makes sense to add this to the repo as well. Since nobody is going to use this code, its only purpose would be to serve as a best-practice example to beginners. However, glossing over the code for a minute, I see a lot of issues that would need to be ironed out to the point where I'd consider rewriting the mapping entirely. 
Please do not misinterpret this as a personal attack, I just don't think it makes sense to merge this in its current state.

ywwg committed:
MiniMixxx: Apply eslint

ywwg committed:
MiniMixxx: cleanup and fix comments

ronso0 commented:
Maybe mentioning it in the changelog draws attention to it.
But maybe a blog post would be better? "open source hardware & software"

Be-ing commented:
I agree, it doesn't make much sense to ship this with Mixxx if the device isn't readily commercially available. If you want to post it on the forum and tell users how they can order their own controller like yours that would be cool though.

Swiftb0y reviewed (changes requested):
I said I didn't want to spend and time of this, but I made an exception because I just wanted to point on some high level stuff I noticed which are probably based on incorrect assumptions about the current state of the controller mapping capabilities.

This mapping would've probably benefited a lot from our Components Library. You essentially re-implemented a version of it with the baseclass being called Mode instead of Component.

ywwg commented:
lol in zulip I mentioned it being weird to commit and people said to go ahead.  closing if people don't care

ywwg committed:
MiniMixxx: Upgrade to javascript classes

ywwg committed:
MiniMixx: fix sync button push-and-lock

ywwg committed:
MiniMixxx: Use newer engine connection API

ywwg committed:
Merge branch 'main' of github.com:mixxxdj/mixxx into minimixxx

ywwg committed:
MiniMixxx: use proper debug output

Holzhaus commented:
> I agree, it doesn't make much sense to ship this with Mixxx if the device isn't readily commercially available.

On Zulip, @ywwg mentioned that you *can* buy this controller. I didn't know that either.

The page currently doesn't work, but I suspect it will be fixed at some point:
https://yaeltex.com/product/minimixxx/

So this was a misunderstanding. @ywwg Do you want to reopen this PR?

ywwg committed:
Minimixxx: make it possible to expand tree items in library browsing

ywwg committed:
MiniMixxx: bind booth volume adjustment

Also create mode where pitch sliders don't do anything *unless* shift is held.
This prevents mixtakes

ywwg committed:
minimixxx: adjust jog sensitivity

ywwg committed:
MiniMixxx: make righthand library encoder scroll horizontally

also some minor tweaks

ywwg committed:
Merge branch 'main' into minimixxx

ywwg committed:
Minimixxx: fix typo

ywwg committed:
raspi: make hotcues the non-shift action

ywwg committed:
Merge branch 'main' into minimixxx

ywwg commented:
It's available now! https://factory.yaeltex.com/gallery/show/16088

ywwg commented:
I need to do some cleanup before reposting -- I have some defaults set that most people probably won't want

ywwg committed:
Minimixxx: rename files to be consistent

ywwg commented:
Updated.

As for inventing my own library instead of using the common library, I think it would require an entire rewrite of this config to use the midi-components. And I'm not even sure it would work -- the library assumes each button has a specific function (maybe a shift function) whereas each physical button on this controller may have a number of different functions depending on the mode.    So for instance, whether a button should be lit depends on what mode the button is in. I'm also not sure what advantage I would gain by doing all that work -- the config works quite nicely.

ywwg commented on res/controllers/MiniMixxx-scripts.js:
```diff
@@ -0,0 +1,1410 @@
+    this.loopColor = MiniMixxx.LoopColor;
+    this.color = this.baseColor;
+
+    engine.connectControl(this.channel, "track_loaded", MiniMixxx.bind(MiniMixxx.EncoderModeJog.prototype.trackLoadedHandler, this));
```
done

ywwg commented on res/controllers/MiniMixxx-scripts.js:
```diff
@@ -0,0 +1,1410 @@
+};
+
+// An Encoder represents a single encoder knob and tracks the active mode.
+MiniMixxx.Encoder = function(channel, idx, layerConfig) {
```
done

ywwg commented on res/controllers/MiniMixxx-scripts.js:
```diff
@@ -0,0 +1,1410 @@
+// Set to true to output debug messages and debug light outputs.
+MiniMixxx.DebugMode = false;
+
+// Mixxx's javascript doesn't support .bind natively, so here's a simple version.
```
done

ywwg commented on .eslintrc.json:
```diff
@@ -65,6 +65,7 @@
     "HIDPacket": "writable",
-    "controller": "writable"
+    "controller": "writable",
+    "print": "writable"
```
done

Holzhaus commented on res/controllers/Yaeltex MiniMixxx.midi.xml:
```diff
@@ -0,0 +1,446 @@
+  <info>
+    <name>Yaeltex MiniMixxx</name>
+    <author>Owen Williams</author>
+    <description>MIDI Mapping for the Yaeltex MiniMixx</description>
```
Officially sanctioned by the Redundancy Department of Redundancy. 😉

The description be informative and not just repeat the name, e.g.:

```suggestion
    <description>A compact DJ controller specifically designed for Mixxx.</description>
```

Holzhaus commented on res/controllers/Yaeltex-MiniMixxx-scripts.js:
```diff
@@ -0,0 +1,1511 @@
+
+        for (const layerName in layerConfig) {
+            const mode = layerConfig[layerName];
+            if (mode === "JOG") {
```
Why not use an enum and a switch statement instead of string comparison?

ywwg replied:
oh hah I didn't know js could do that.  I'll fix

ywwg replied:
done

Be-ing commented on res/controllers/Yaeltex-MiniMixxx-scripts.js:
```diff
@@ -0,0 +1,1511 @@
+        this.layers = {};
+        this.activeMode = {};
+
+        for (const layerName in layerConfig) {
+            const mode = layerConfig[layerName];
```
```suggestion
        for (const mode of layerConfig) {
```

ywwg committed:
Minimixxx: use switches where possible

Swiftb0y commented:
> I'm also not sure what advantage I would gain by doing all that work -- the config works quite nicely.

Not much, I'm just personally worried about increased maintenance work with this extra code. If you assure that you will maintain this mapping I don't really have a problem with your custom library, though in the future, I'd prefer if we tried to reiterate on the existing design instead of inventing the wheel from scratch. I'm aware that the current approach to switching functionality within a component is not ideal.

Be-ing commented:
I share @Swiftb0y's concern about writing an ad hoc JavaScript library for this one mapping that serves the same purpose of the library we already have; this seems like a not-invented-here issue. You don't have to rewrite a fully working PR from scratch, but I took a brief look at the code but quickly got discouraged from reviewing it. I don't want to bother learning a completely different system for every mapping and I think it's asking a lot of reviewers to do so. If you have an issue with the Components library, we can discuss that, especially as we are making backwards incompatible changes in #4171.

ywwg commented:
yeah next time I would try using the existing lib.  And yes, I will maintain the mapping.

ywwg committed:
Update res/controllers/Yaeltex MiniMixxx.midi.xml

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

ywwg committed:
Update res/controllers/Yaeltex-MiniMixxx-scripts.js

Co-authored-by: Be <be.0@gmx.com>

ywwg commented:
documentation in progress: https://github.com/mixxxdj/manual/pull/456

Swiftb0y commented:
Please take care of the pre-commit errors before I'll start to review (silence them if absolutely necessary).

ywwg committed:
Merge remote-tracking branch 'origin/main' into minimixxx

ywwg committed:
Minimixxx: revert bad change (var was undefined)

ywwg commented:
@Swiftb0y errors fixed

Swiftb0y commented:
Thank you. I'll review shortly, feel free to ping if I forget to do it in the next 7 days.

Swiftb0y reviewed (commented):
Code looks good. Had some minor nitpicks but nothing worth leaving a comment for.
However, I'd propose to add a disclaimer at the top though to discourage beginners to copy paste this code into their own mappings. Most of the code relies on the architecture of the mapping with your custom library, so users would need to copy that library as well, which we don't want since we agreed that it would not be maintainable to introduce another library we have to support. Is that ok for you?

ywwg commented:
yeah no problem -- I can add a comment that people should use the component library if they are looking for copypasta.

ywwg committed:
Minimixxx: disclaimer

Swiftb0y reviewed (approved):
Thanks

Holzhaus commented:
@Swiftb0y we should not merge new mappings without also merging the corresponding manual PR (mixxxdj/mixxx#456 in this case).

Swiftb0y commented:
I'm sorry, forgot about that. Doesn't make much sense to revert the PR now, right?

Holzhaus commented:
> I'm sorry, forgot about that. Doesn't make much sense to revert the PR now, right? 

No, let's just try to remember it next time.😅

daschuer commented:
@Swiftb0y: Please create a merge commit next time. Else it is hard to track changes among all the "Merge pull request #4521 from ... " messages. Also the individual commit messages are loosing the context. 
In case you like to eliminate faulty commits, rebase the branch before merge.

Swiftb0y commented:
We need to write down proper guidelines for this. This has been more than once that I merged too soon or merged in the wrong mode. As a result, I would simply not press merge anymore if I have to take the blame for each little mistake.

Holzhaus commented:
I think the "let's merge mappings when both  documentation and code are ready" is known. You forgot it this time, no big issue, just try to remember it next time ;-)

Regarding merge commit vs. squash: I disagree with @daschuer and I think squash+merge was the right call, the commit messages were not that helpful anyway.

Swiftb0y commented:
> I think the "let's merge mappings when both documentation and code are ready" is known. 

Yes that is known, though writing it down into a checklist would still make sense so I can take a look before merging to make sure I didn't forget anything.

daschuer commented:
@Holzhaus In this case it is not worth to discuss it. But let's clarify our approach for the future: 

Please have a look at the commit message here: 
https://github.com/mixxxdj/mixxx/commit/c0679a2839480e058bd1670942294f9dfb9a7c29

The original commit messages are still there, but useless because they are decoupled from the original commit. 

The other issue is the produced git history:  
I am currently crawling down the commit history from merge commit to merge commit by searching for:
"Merge pull request #4545" to write the CHANGELOG (still 190 Merges left :-/ ) 
I need to be careful not to skip the squashed merges. If we now consider sqash merges with edited commit message we will have hard times to follow the original main branch and not branch into a PR. 
A main branch with only merge commits is a better alternative here.  
 
That are the reasons for our rule in https://github.com/mixxxdj/mixxx/blob/main/CONTRIBUTING.md
``
Merge PRs using a merge , to keep the original commits valid. Keep the default commit message "Merge pull request ..." with the reference to the pull request. In case where the PR contains broken (non-building) commits, back-and-forth commits or commits without a meaningful commit message that are not worth keeping, ask the author to squash the commits before merge or use squash-and-merge if the author is not fluent in git. 
``

In this case the commit messages where not so much cluttered to be useless so I would have preferred to just merge them as they are. If you really want to go the extra mile for a cleaner history, which is IMHO not worth the effort. 
You could have ask @ywwg to clean up the history for more reasonable commit messages and then merge. 

In general, I prefer to put important info as comment into the code, this is a better invest of time.

Does this sound reasonable? Do we need an adjustment of our rules?

Holzhaus commented:
> The original commit messages are still there, but useless because they are decoupled from the original commit. 

I usually edit the commit message for the squash commit to something like "Add MiniMixxx controller mapping" 😉

> I am currently crawling down the commit history from merge commit to merge commit

Why? For the changelog we have the "changelog" label now. PRs without that label should not be mentioned in the changelog.

> In this case the commit messages where not so much cluttered to be useless so I would have preferred to just merge them as they are. If you really want to go the extra mile for a cleaner history, which is IMHO not worth the effort. 
> You could have ask @ywwg to clean up the history for more reasonable commit messages and then merge. 

In this case we could do that, but most mapping contributors are not fluent in git, and IMHO the bar to get a mapping g merged is already pretty high. We try to have good code/documentation quality without scaring away contributors. For that reason I think squash+merge is preferable over asking mapping contributors to do  squash/rebase manually. 

> In general, I prefer to put important info as comment into the code, this is a better invest of time.

Yes.

daschuer commented:
Ok, fine, we are basically on the same line. 

So the mayor difference between our opinion is when the squash+merge is preferred. 

In terms of the lowest bar, I prefer just merge. There is no harm with an extra commit like "MiniMixxx: cleanup" or such. 
If you insist to squash, you need to write a new commit message that describes the new commit. If we want to keep the single commits messages in a list like it was done here, we should also keep the commits. 

Can we agree to that? 
  

Regarding our Changelog lable, I do not really trust it. Can this replace a systematic approach looking at all PRs? 
Even though if we do remember all the time to set the flag, you do not see the final picture when writing the CHANGELOG. you need to decide which one is worth to not and which not. 

For now I continue for 2.4 with the style already started: https://github.com/mixxxdj/mixxx/blob/main/CHANGELOG.md

Is this OK?
